### PR TITLE
Implement division in range analysis

### PR DESCRIPTION
**Pull request by tnkhanh**

### CODE REVIEW

File: src/dataflow/range.ml [hunk 1]
```diff
@@ -208,6 +208,11 @@ module IntervalDomain = struct
     add_bound a neg_b
   ;;
 
+  let udiv_bound (a : bound) (b : bound) : bound =
+    (*TO DO *)
```
Comment (taquangtrung) [inline src/dataflow/range.ml:212]: (* TODO *)

File: src/dataflow/range.ml [hunk 2]
```diff
@@ -297,6 +302,13 @@ module IntervalDomain = struct
     mk_range lb ub ~li ~ui
   ;;
 
+  let udiv_range (a : range) (b : range) =
+    (* TO TEST *)
+    let lb = udiv_bound a.range_lb b.range_lb in
+    let ub = udiv_bound a.range_ub b.range_ub in
```
Comment (taquangtrung) [inline src/dataflow/range.ml:308]: This logic is INCORRECT!

**Review by taquangtrung — CHANGES_REQUESTED** [review]

File: src/dataflow/range.ml (hunk 2 above)
Comment (tnkhanh) [reply src/dataflow/range.ml:308]: Ah, let me think about it.

Comment (taquangtrung) [reply src/dataflow/range.ml:308]: @tnkhanh : you need to put effort into the work. Cannot just randomly implement the code like that!

**Comment by tnkhanh** [comment]
My logic to implement next is like this:
If a, b are both positive (unsigned), then `ub(a/b) = ub(a) / lb(b) , lb(a/b) = lb(a) / ub(b)`. This looks simple and I will implement it first
For signed integers, I will divide into four cases: a is positive/negative and b is positive/negative
We also need add Division by zero later.

**Comment by taquangtrung** [comment]
@tnkhanh: need to think of a general solution for both positive/negative integers.
The algorithm for only positive integers might NOT work for all integers.

**Comment by taquangtrung** [comment]
@tnkhanh: need to put more effort into the implementation!

**Comment by tnkhanh** [comment]
I'm working on implementing `Division-by-zero` bug detection.
Initially, I wanted it to be detected  by `check_bug_integer_underflow` and `check_bug_integer_overflow` which is not ideal. This is also difficult because these functions only return one `bug option` when called, they can't both report an `IntegerOverflow` and a `DivisionByZero`
So, my idea now is to implement `check_bug_division_by_zero`. The `check_bug_integer_overflow` will skip the `DivisionByZero` bug and this bug is only detected by `check_bug_division_by_zero`. `check_bug_division_by_zero` will check the range of every divisors to see if it contains 0. 
Is this feasible? @taquangtrung

**Comment by taquangtrung** [comment]
Yes, separating division_by_zero into a separate bug pass is reasonable.

Do note that the division_by_zero reported by range analysis may contain many false positives (since the range may contain 0, but the actual divisor might not be 0), so need to add a note for future improvement to reduce false positives.

File: src/dataflow/range.ml [hunk 3]
```diff
@@ -208,6 +208,57 @@ module IntervalDomain = struct
...
+    | NInf, NInf -> error "sdiv_bound: undefined for NInf / NInf"
+    | _, PInf | _, NInf -> BInt BInt.zero
+    | PInf, _ ->
+      let cmp = compare_bound b (BInt BInt.zero) in
+      if cmp > 0 then PInf
+      else
+      if cmp < 0 then NInf
+      else error "sdiv_bound: undefined for PInf / 0"
+    | NInf, _ ->
+      let cmp = compare_bound b (BInt BInt.zero) in
```
Comment (taquangtrung) [inline src/dataflow/range.ml:245]: Use `Int64.zero`

File: src/dataflow/range.ml [hunk 4]
```diff
@@ -281,6 +332,24 @@ module IntervalDomain = struct
     | _, BInt x -> mult_bint_bound x a
   ;;
 
+  let union_range (a : range) (b : range) : range =
```
Comment (taquangtrung) [inline src/dataflow/range.ml:335]: Is it still needed?

File: src/dataflow/range.ml [hunk 5]
```diff
@@ -297,6 +366,119 @@ module IntervalDomain = struct
...
+  (*both a and b are only-positive or only-negative *)
+  let sdiv_range_one_sign (a : range) (b : range) : range = 
+    let max_bound b1 b2 =
+      if compare_bound b1 b2 > 0
+      then b1
+      else b2 in
+    let min_bound b1 b2 =
+      if compare_bound b1 b2 < 0
+      then b1
+      else b2 in
```
Comment (taquangtrung) [inline src/dataflow/range.ml:401]: Move out as new separate functions

File: src/dataflow/range.ml [hunk 6]
```diff
@@ -297,6 +366,119 @@ module IntervalDomain = struct
...
+        (union_range (range_of_bound bzero) (sdiv_range (pos_int_range a) b) )
+        (sdiv_range (neg_int_range a) b) *)
+    let max_bound b1 b2 =
+      if compare_bound b1 b2 > 0
+      then b1
+      else b2 in
+    let min_bound b1 b2 =
+      if compare_bound b1 b2 < 0
+      then b1
+      else b2 in
```
Comment (taquangtrung) [inline src/dataflow/range.ml:450]: Repeated code, see above comment!

File: src/dataflow/range.ml [hunk 7]
```diff
@@ -297,6 +366,119 @@ module IntervalDomain = struct
...
+              (sdiv_bound a_lb b_one)::(sdiv_bound a_ub b_one)::acc
+            else
+              acc)
+      |> (fun acc ->
+            if (compare_bound b_minus_one b_lb >=0 && 
+                compare_bound b_minus_one b_ub <=0) 
+            then
+              (sdiv_bound a_lb b_minus_one)::(sdiv_bound a_ub b_minus_one)::acc
+            else
+              acc) in
```
Comment (taquangtrung) [inline src/dataflow/range.ml:475]: Can improve this part? Look ugly!

**Review by taquangtrung — CHANGES_REQUESTED** [review]

File: src/dataflow/range.ml (hunk 2 above)
Comment (tnkhanh) [reply src/dataflow/range.ml]: fixed

File: src/dataflow/range.ml [hunk 8]
```diff
@@ -208,6 +208,57 @@ module IntervalDomain = struct
...
+  let sdiv_bound (a : bound) (b : bound) : bound =
+    (*TODO: Division by zero *)
+    match a, b with
+    | PInf, PInf -> error "sdiv_bound: undefined for PInf / PInf"
+    | PInf, NInf -> error "sdiv_bound: undefined for PInf / NInf"
+    | NInf, PInf -> error "sdiv_bound: undefined for NInf / PInf"
+    | NInf, NInf -> error "sdiv_bound: undefined for NInf / NInf"
+    | _, PInf | _, NInf -> BInt BInt.zero
+    | PInf, _ ->
+      let cmp = compare_bound b (BInt BInt.zero) in
```
[the remark this answers is not in the document]
Comment (tnkhanh) [reply src/dataflow/range.ml]: Used `Int64`

**Pull request merged**